### PR TITLE
Added searchable encryption functionality to library

### PR DESCRIPTION
**Pull request by AmalieDue**
### Description
Added simple draft searchable encryption functionality to library.

closes DEV-80

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [x] Added technical feature
- [x] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Kubernetes changes
- [ ] CI/CD workflow changes

### Testing Checklist
- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make tests`.
- [ ] I have updated relevant CI/CD workflows.
- [ ] I have updated relevant Kubernetes manifests/scripts.
- [ ] I have updated/added relevant (internal and external) documentation (readme, doc comments, etc).
- [ ] I have updated the dependency map to reflect my changes.
- [ ] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: lib/searchable.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	count, err := i.Count(masterKey, keyword)
+	if err != nil {
+		return err
+	}
+	label, err := tagger.Tag(uint64Encode(count))
+	if err != nil {
+		return err
+	}
+
+	wrappedKey, encryptedID, err := cryptor.Encrypt([]byte(id), 0)
```
Comment (philip-cc) [inline lib/searchable.go:60]: Maybe an empty string makes more sense here:
```suggestion
	wrappedKey, encryptedID, err := cryptor.Encrypt([]byte(id), "")
```

File: lib/searchable.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	decryptedIDs := []string{}
+
+	var count uint64
+	for {
+		label, err := tagger.Tag(uint64Encode(count))
+		if err != nil {
+			return nil, err
+		}
+
+		if encryptedID, ok := i.Mapping[base64.StdEncoding.EncodeToString(label)]; ok {
```
Comment (philip-cc) [inline lib/searchable.go:93]: I'd suggest flipping this check to make the code simpler, e.g.
```
if !ok {
  break
}

// Do decryption stuff
```

File: lib/searchable.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	cryptor, err := crypto.NewAESCryptor(k2)
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedIDs := []string{}
+
+	var count uint64
+	for {
```
Comment (philip-cc) [inline lib/searchable.go:87]: You can have a for-loop with an empty stop condition:
```suggestion
	for count := 0 ; ; count++ {
```

File: lib/searchable.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+func (i *Index) Count(masterKey []byte, keyword string) (uint64, error) {
+	k1 := crypto.KMACKDF(len(masterKey), masterKey, []byte("mac"), []byte(keyword))
+
+	tagger, err := crypto.NewKMAC256Tagger(k1)
+	if err != nil {
+		return 0, err
+	}
+
+	var count uint64
+	for {
```
Comment (philip-cc) [inline lib/searchable.go:118]:
```suggestion
	for count := 0 ; ; count++ {
```

File: lib/searchable_test.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+	for i := 0; i < len(keywords); i++ {
+		for j := 0; j < len(ids); j++ {
+			err = searchable.Add(masterKey, keywords[i], ids[j])
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+	if len(searchable.Mapping) != 4 {
```
Comment (philip-cc) [inline lib/searchable_test.go:69]: To make life easier in the future. 
```suggestion
	if len(searchable.Mapping) != len(keywords)*len(ids) {
```

File: lib/searchable_test.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+		err = searchable.Add(masterKey, keyword, ids[i])
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		IDs, err := searchable.Search(masterKey, keyword)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if IDs[len(IDs)-1] != ids[i] {
```
Comment (philip-cc) [inline lib/searchable_test.go:96]: We should probably compare `IDs` to `ids[:i]` instead. That would also make the next test redundant.

File: lib/searchable.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		} else {
+			break
+		}
+		count++
+	}
+
+	return decryptedIDs, nil
+}
+
+func (i *Index) Count(masterKey []byte, keyword string) (uint64, error) {
```
Comment (philip-cc) [inline lib/searchable.go:109]: Does this need to be exported?

File: lib/searchable_test.go [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+func TestCount(t *testing.T) {
+	searchable := NewSearchable()
+
+	rand := &crypto.NativeRandom{}
+	masterKey, err := rand.GetBytes(32)
+	if err != nil {
+		t.Fatalf("Random failed: %v", err)
+	}
+
+	keyword := "first keyword"
```
Comment (philip-cc) [inline lib/searchable_test.go:214]: We should probably test this with more than one keyword present.

File: lib/searchable.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+package encryptonize
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"encryptonize/crypto"
+)
+
+type Index struct {
+	Mapping map[string]SealedID
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+type SealedID struct {
+	Ciphertext []byte
+	WrappedKey []byte
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+package encryptonize
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"encryptonize/crypto"
+)
+
+type Index struct {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/searchable.go:26]: Fixed

File: lib/searchable.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+type SealedID struct {
+	Ciphertext []byte
+	WrappedKey []byte
+}
+
+func NewSearchable() *Index {
+	mapping := make(map[string]SealedID)
+
+	return &Index{Mapping: mapping}
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+func (i *Index) Add(masterKey []byte, keyword, id string) error {
+	k1 := crypto.KMACKDF(len(masterKey), masterKey, []byte("mac"), []byte(keyword))
+	k2 := crypto.KMACKDF(len(masterKey), masterKey, []byte("id"), []byte(keyword))
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	WrappedKey []byte
+}
+
+func NewSearchable() *Index {
+	mapping := make(map[string]SealedID)
+
+	return &Index{Mapping: mapping}
+}
+
+func (i *Index) Add(masterKey []byte, keyword, id string) error {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable.go (hunk 2 above)
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable_test.go (hunk 6 above)
Comment (AmalieDue) [reply lib/searchable_test.go]: Fixed

File: lib/searchable.go (hunk 7 above)
Comment (AmalieDue) [reply lib/searchable.go]: Fixed

File: lib/searchable_test.go (hunk 8 above)
Comment (AmalieDue) [reply lib/searchable_test.go]: Fixed

**Comment by AmalieDue** [comment]
I have fixed all the comments. The tests fail, but they are meant to, due to the change of repositories.

**Comment by philip-cc** [comment]
Repalced by https://github.com/cyber-crypt-com/encryptonize-lib/pull/2

**Pull request closed without merging**